tex2pdf_service/tex2pdf/doc_converter.py

```diff
@@ -50,41 +50,37 @@ def combine_documents(doc_list: typing.List[str], out_dir: str, out_filename: st
-                        pdf.pages.extend(pdf_page.pages)
+    
+    logger = get_logger()
+    effective_pdf_list = []
+    # first collection list of pdfs to be combined (normal and converted images)
+    for doc_path in doc_list:
+        [stem, ext] = os.path.splitext(doc_path)
+        # This should exist but be safe.
+        if not os.path.exists(doc_path):
+            continue
+        if ext.lower() == ".pdf":  # This should not need lower() but be safe. Should I assert?
+            effective_pdf_list.append(doc_path)
+        elif ext.lower() in graphics_exts:
+            temp_pdf = os.path.join(out_dir, stem + '.pdf')
+            try:
+                pdf_filename = convert_image_to_pdf(doc_path, temp_pdf)
+                if pdf_filename and os.path.exists(pdf_filename):
+                    effective_pdf_list.append(temp_pdf)
                     converted_docs.append(doc_path)
-                except PdfError:
-                    failed_docs.append(doc_path)
-                    logger.warning("Cannot open PDF file %s", doc_path, extra=log_extra)
-                    pass
-            elif ext.lower() in graphics_exts:
-                temp_pdf = os.path.join(out_dir, stem + '.pdf')
-                try:
-                    pdf_filename = convert_image_to_pdf(doc_path, temp_pdf)
-                    if pdf_filename and os.path.exists(pdf_filename):
-                        converted_docs.append(doc_path)
-                        with pikepdf.Pdf.open(pdf_filename) as pdf_page:
-                            pdf.pages.extend(pdf_page.pages)
-                        pass
-                except UnidentifiedImageError:
-                    failed_docs.append(doc_path)
-                    logger.warning("Unsupported %s", doc_path, extra=log_extra)
-                    pass
-                except Exception as _exc:
-                    failed_docs.append(doc_path)
-                    logger.warning("Unknown error %s", doc_path, extra=log_extra,
-                                   exc_info=True)
-                    pass
-                pass
-            pass
-        pdf.save(output_path)
+            except UnidentifiedImageError:
+                failed_docs.append(doc_path)
+                logger.warning("Unsupported %s", doc_path, extra=log_extra)
+            except Exception as _exc:
+                failed_docs.append(doc_path)
+                logger.warning("Unknown error %s", doc_path, extra=log_extra,
+                               exc_info=True)
+    # call gs to combine the pdf
+    # we cannot use pikepdf (easily) here since it breaks annotations (links)
+    gs_cmd = [
+        "gs", "-sDEVICE=pdfwrite", "-dNOPAUSE", "-dBATCH", "-dSAFER", f"-sOutputFile={output_path}"
+    ] + effective_pdf_list
+    logger.debug("Running gs to combine pdfs: %s", shlex.join(gs_cmd), extra=log_extra)
+    # exception handing is done in convert_driver:_finalize_pdf
+    subprocess.run(gs_cmd, timeout=30, check=True)
```

Comment:
No need for error checking? This isn't very visible if anything goes wrong. Not even a warning logging.

Reply:
Error checking is done via exceptions, due to `check = True` and the exception handing in the convert driver itself:
```
        except subprocess.TimeoutExpired:
            logger.warning("Failed combining PDFs: gs call timed out", extra=self.log_extra)
        except subprocess.CalledProcessError:
            logger.warning("Failed combining PDFs: gs returned an error", extra=self.log_extra)
```

This is in line with the other exception handling for pdf open errors etc that might throw exceptions in the process, but are treated in the `_finalize_pdf` method of convert driver.

So errors are logged.